---
title: "FDR Exercises"
author: "Stephen Blatti"
date: "September 3, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

In this assessment we will define error controlling procedures for experimental data. We will make list of genes based on q-values. We will also assess your understanding of false positives rates and false negative rates by asking you to create a Monte Carlo simulation.

You will need to install the following libraries:

```{r}
library(devtools)
library(rafalib)
# install_github("genomicsclass/GSE5859Subset")
# install_bioc("genefilter")
# install_bioc("qvalue")
```
1. Load the gene expression data
```{r}
library(GSE5859Subset)
data(GSE5859Subset)
```
We are interested in comparing gene expression between the two groups defined in the sampleInfo table.

Compute a p-value for each gene using the function rowttests from the genefilter package in Bioconductor.
```{r}
library(genefilter)
?rowttests
```
How many genes have p-values smaller than 0.05?
```{r}
g = sampleInfo$group #Get the group index from sampleInfo
ttests = rowttests(geneExpression, factor(g))
pvals = ttests$p.value
alpha = 0.05
sum(pvals < alpha)
```
2. Apply the Bonferroni correction to the p-values obtained in question #1 to achieve a FWER of 0.05. How many genes are called significant under this procedure?

```{r}
m <- length(pvals)
k <- alpha / m
sum(pvals < k)
```
3. Note that the FDR is a property of a list of features, not each specific feature. The q-value relates FDR to an individual feature. To define the q-value we order features we tested by p-value then compute the FDRs for a list with the most significant, the two most significant, the three most significant, etc... The FDR of the list with the, say, m most significant tests is defined as the q-value of the m-th most significant feature. In other words, the q-value of a feature, is the FDR of the biggest list that includes that gene.

In R, we can compute the q-value using the p.adjust function with the FDR option. Read the help file for p.adjust and then, for our gene expression dataset, compute how many genes achieve an FDR < 0.05
```{r}
#?p.adjust
g <- factor(sampleInfo$group)
pvals <- rowttests(geneExpression,g)$p.value
fdr_aveQ <- p.adjust(pvals,method="fdr")
sum(fdr_aveQ < alpha)
```
      Note that controlling the FDR at 0.05 gives us 3 more genes than the Bonferroni correction. Note that we are controlloing two very different error rates. Here we are saying that we think this list of 13 genes has about 5% false positives. The Bonferroni procedure gave us a list ot 10 genes for which we were quite certain had no false positives. Note again that we have not discussed false negatives.

4. Now use the qvalue function, in the Bioconductor qvalue package, to estimate q-values using the procedure described by Storey.

Using this estimate how many genes have q-values below 0.05?

```{r}
library(qvalue)
?qvalue
res <- qvalue(pvals)
sum(res$qvalues < alpha)
```
        Now the list has increased to 17. However, we are also claiming this list has an FDR of 5%. The reason this list is different is because as explained in the videos, qvalue, estimates FDR differently and is less conservative. Remember that the theory provides bounds for FDR: it guarantees FDR will be less than 0.05. If qvalue does in fact estimate pi0 well then it will provide a list with FDR closer to 0.05. 

5. Read the help file for qvalue and report the estimated proportion of genes for which the null hypothesis is true pi_0 = m_0 / m.
```{r}
res$pi0
# or qvalue(pvals)$pi0
```
6. Note that we have the number of genes passing the q-value <0.05 threshold is larger with the qvalue function than the p.adjust difference.
Why is this the case? Make a plot of the ratio of these two estimates to help answer the question.

    One of the two procedures is flawed.
    The two functions are estimating different things.
    The qvalue function estimates the proportion of genes for which the null hypothesis is true and provides a less conservative estimate Correct
    The qvalue function estimates the proportion of genes for which the null hypothesis is true and provides a more conservative estimate  

```{r}
plot(qvalue(pvals)$qvalue/p.adjust(pvals,method="fdr"))
abline(h=qvalue(pvals)$pi0,col=2)

# To get an idea of how pi0 is estimated, note that if we look at the histogram, pi0 roughly tells us the proportion that looks about uniform:
  
hist(pvals,breaks=seq(0,1,len=21))
expectedfreq <- length(pvals)/20 #per bin
abline(h=expectedfreq*qvalue(pvals)$pi0,col=2,lty=2)
```

7. Note that this is an advanced question and that you can ask questions in the discussion forum.

Create a Monte Carlo Simulation in which you simulate measurements from 8,793 genes for 24 samples: 12 cases and 12 controls.
```{r}
# n <- 24
# m <- 8793
# mat <- matrix(rnorm(n*m),m,n)
```
Now for 500 genes, there is a difference of 2 between cases and controls:
```{r}
# delta <- 2
# positives <- 500
# mat[1:positives,1:(n/2)] <- mat[1:positives,1:(n/2)]+delta
```
So the null hypothesis is true for 8793-500 genes. Using the notation from the videos m=8793, m0=8293 and m1=500

Set the seed at 1, set.seed(1) and run this experiment 1,000 times with a Monte Carlo simulation. For each instance compute p-values using a t-test (using rowttests in the genefilter package) and create three lists of genes using:

    Bonferroni correction to achieve an FWER of 0.05,
    p-adjust estimates of FDR to achieve an FDR of 0.05, and
    qvalue estimates of FDR to to achieve an FDR of 0.05.

For each of these three lists compute the number of false positives in the list and the number of false negatives: genes not in the list that should have been because the null hypothesis is not true (we added 2 to the controls to create the cases).

What is the false positive rate (false positives divided by m0) if we use Bonferroni?
```{r}
n <- 24
m <- 8793
B <- 1000
delta <-2
positives <- 500
m0 <- m - positives
m1 <- positives
g <- factor( rep(c(0,1),each=12) )
set.seed(1)
mc_sim <- replicate(B,{
  mat <- matrix(rnorm(n*m),m,n)
  mat[1:positives,1:(n/2)] <- mat[1:positives,1:(n/2)] + delta
  pvals = rowttests(mat,g)$p.val
  # Bonferroni
  FP1 <- sum(pvals[-(1:positives)] <= 0.05 / m)
  FN1 <- sum(pvals[1:positives] > 0.05 / m)
  # p.adjust q-value
  qvals1 = p.adjust(pvals,method="fdr")
  FP2 <- sum(qvals1[-(1:positives)] <= 0.05)
  FN2 <- sum(qvals1[1:positives] > 0.05)
  # qvalue q-value
  qvals2 = qvalue(pvals)$qvalue
  FP3 <- sum(qvals2[-(1:positives)] <= 0.05)
  FN3 <- sum(qvals2[1:positives] > 0.05)  
  c(FP1,FN1,FP2,FN2,FP3,FN3)
  })
mean(mc_sim[1,] / m0) 
```

    Note that this value is much smaller than 0.05. This is because Bonferroni controls FWER to be 0.05 not the FDR. In this case, controlling FWER to be 0.05 gives us very low FDR. This makes intuitive sense since having just 1 mistake out of 8,293 possible mistakes is very small and we trying to avoid even 1. 

8. From the same Monte Carlo simulation as in the question above, what is the false negative rate if we use Bonferroni?
```{r}
mean(mc_sim[2,] / m1)
```

    Note that having a very low FDR comes at a cost. Namely that we increase our false negative rate, in this case to 76%. This means we miss including the great majority of genes for which the null is not true. This trade-off is always present when we have to pick a cutoff. Understanding the tradeoff can help us determine what appraoch is better in the context of our scientific problem.
    
9. From the same Monte Carlo simulation as in question #7, what is the false positive rate if we use q-values from p.adjust?

```{r}
mean(mc_sim[3,] / m0) 
```

    Note that although much higher than the FDR for Bonferroni, the FDR is substantially lower than 0.05 we were shooting for. This is because the Benjaminiâ€“Hochberg procedure gives us a bound. The larger m1, the more conservative this approximation will be.

10. From the same Monte Carlo simulation as in question #7, what is the false negative rate if we use q-values from p.adjust?

```{r}
mean(mc_sim[4,] / m1) 
```

    Here we see the potential advantage of FDR over FWER, in particular if our goal is discovery. The false negative rate is much reduced now from 0.76 to 0.08

11. From the same Monte Carlo simulation as in question #7, what is the false positive rate if we use q-values from qvalue function?
```{r}
mean(mc_sim[5,] / m0) 
```

    Here we see that by estimating pi0 this approach gets closer to the targeted FDR of 0.05.

12. From the same Monte Carlo simulation as in question #7, what is the false negative rate if we use q-values from qvalue function?

```{r}
mean(mc_sim[6,] / m1) 
```

    Here we see that by creating a list of an FDR closer to 0.05 we are less conservative and thus decrease the false negative rate further.
